File: 110. Balanced Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def checkheight(self,root):
        if root==None:
            return 0
        return max(self.checkheight(root.left),self.checkheight(root.right))+1
    def isBalanced(self, root: TreeNode) -> bool:
        if root==None:
            return True
        le=self.checkheight(root.left)
        rh=self.checkheight(root.right)
        if le-rh>1 or le-rh<-1:
            return False
        
        return self.isBalanced(root.left) and self.isBalanced(root.right)

File: 110_Balanced_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int caldepth(TreeNode* root)
    {
        if (root==NULL)
        {
            return 1;
        }
        
        return 1+max(this->caldepth(root->right),this->caldepth(root->left));
    }
    bool isBalanced(TreeNode* root) {
        if (root==NULL)
        {
            return true;
        }
        int left=this->caldepth(root->left);
        int right=this->caldepth(root->right);
        int sum=abs(right-left);
        return ((sum<2) && this->isBalanced(root->left) && this->isBalanced(root->right));
    }
};